storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1666 @@ buf_block_t *fsp_get_latched_xdes_page(
+7) Add the fseg_not_full extent to fseg_full list
+
+Above all scenario done by 2 steps.
+1) prepare - Basically validates the necessary condition
+and make sure that pages are being latched
+2) Complete - Completes the action by using the latched
+pages in prepare step */
+class pageAllocator
+{
+private:
+  /* System tablespace */
+  fil_space_t *m_space=  fil_system.sys_space;
+  /* Header block for the tablespace */
+  buf_block_t *m_header_block= nullptr;
+  /* Index node block */
+  buf_block_t *m_iblock= nullptr;
+  /* Index node */
+  fseg_inode_t *m_inode= nullptr;
+  /* offset of index node within index node page*/
+  uint16_t m_ioffset= 0;
+  /* Maximum free limit of the tablespace */
+  uint32_t m_free_limit= 0;
+  /* Segment id */
+  uint64_t m_seg_id= 0;
+  /* Extent size */
+  uint32_t m_extent_size= 0;
+
+  /* New page */
+  uint32_t m_new_page= 0;
+  /* New block to be allocated */
+  buf_block_t *m_new_block= nullptr;
+  /* New block extent descriptor */
+  buf_block_t *m_new_xdes= nullptr;
+  /* New block descriptor */
+  xdes_t *m_new_descr= nullptr;
+  /* New block descriptor offset within xdes page */
+  uint16_t m_xoffset= 0;
+  /* New extent descriptor state */
+  uint32_t m_new_state= 0;
+  /* Need segment allocation */
+  bool m_need_segment= false;
+  /* Old pages during allocation to be saved */
+  std::map<uint32_t, buf_block_t*> m_old_pages;
+  /* Mini-transaction to allocate a page */
+  mtr_t *m_mtr;
+
+  /** Save the old page state of the block before
+  allocating a page
+  @param block   block to be stored
+  @return error code */
+  dberr_t save_old_page(buf_block_t *block) noexcept
+  {
+    if (!block) return DB_SUCCESS;
+    uint32_t page_no= page_get_page_no(block->page.frame);
+    if (m_old_pages.find(page_no) != m_old_pages.end())
+      return DB_SUCCESS;
+
+    buf_block_t *old= buf_LRU_get_free_block(have_no_mutex_soft);
+    if (!old) return DB_OUT_OF_MEMORY;
+    memcpy_aligned<UNIV_PAGE_SIZE_MIN>(
+      old->page.frame, block->page.frame, srv_page_size);
+
+    m_old_pages[page_no]= old;
+    return DB_SUCCESS;
+  }
+
+  /** Prepare the free extent allocation by validating
+  FLST_PREV, FLST_NEXT of choosen extent descriptor
+  and their FLST_LEN of FSP_FREE list in FSP_HEADER_PAGE
+  @return error code or DB_SUCCESS */
+  dberr_t free_extent_prepare() noexcept
+  {
+    /* At least there should be 1 element in FSP_FREE list */
+    byte *len=
+      &m_header_block->page.frame[FSP_HEADER_OFFSET + FSP_FREE
+                                  + FLST_LEN];
+    if (mach_read_from_4(len) == 0)
+      return DB_CORRUPTION;
+
+    buf_block_t *prev_free_lst= nullptr;
+    buf_block_t *next_free_lst= nullptr;
+    dberr_t err= flst::remove_prepare(m_new_descr, m_free_limit,
+                                      &prev_free_lst,
+                                      &next_free_lst, m_mtr);
+    if (err == DB_SUCCESS)
+    {
+      save_old_page(next_free_lst);
+      save_old_page(prev_free_lst);
+    }
```

Comment:
Can `remove_complete()` and `add_last_complete()` operate on the same list under any circumstances? I think that a source code comment that explains this would be very helpful here.

If any overlap is possible, then it would seem to me that we need to complete one of the `remove` or `add_last` before preparing the other, so that if one of the operations is exposing a corruption to the other, we’d be able to roll back the whole.